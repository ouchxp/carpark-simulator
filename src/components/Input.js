import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import styled from 'styled-components';
import { EXAMPLE_COMMANDS } from '../const';
import { startCommandSequence } from '../sagas/actions';

const InputArea = styled.textarea`
  font-family: monospace;
  width: -moz-available;
  width: -webkit-fill-available;
  height: 140px;
  resize: none;
  border: 1px solid lightgrey;
  border-radius: 4px;
`;

const ExecuteButton = styled.button`
  display: block;
  margin-right: 0;
  margin-left: auto;
`;

const Input = ({ isRunning, dispatch }) => {
  const inputRef = useRef(null);

  const executeCommands = () => {
    const commands = inputRef.current.value.trim().split(/\r?\n/);
    dispatch(startCommandSequence(commands));
  };

  return (
    <>
      <label htmlFor="commandInput">Input:</label>
      <InputArea
        id="commandInput"
        ref={inputRef}
        defaultValue={EXAMPLE_COMMANDS.trim()}
        placeholder="write one command each line"
        disabled={isRunning}
      />
      <ExecuteButton onClick={executeCommands} disabled={isRunning}>
        Execute
      </ExecuteButton>
    </>
  );
};

Input.propTypes = {
  isRunning: PropTypes.bool.isRequired,
  dispatch: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  isRunning: state.app.isRunning,
});

export default connect(mapStateToProps)(Input);
