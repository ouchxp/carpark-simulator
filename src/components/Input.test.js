import React from 'react';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import Input from './Input';

const mockStore = configureStore([]);

describe('Input', () => {
  it('disable the execute button and textarea when running', () => {
    const store = mockStore({
      app: { isRunning: true },
    });
    const wrapper = mount(
      <Provider store={store}>
        <Input />
      </Provider>
    );
    expect(wrapper.find('InputArea').get(0).props.disabled).toBe(true);
    expect(wrapper.find('ExecuteButton').get(0).props.disabled).toBe(true);
  });

  it('dispatch startCommandSequence action when button clicked', () => {
    const store = mockStore({
      app: { isRunning: false },
    });
    const wrapper = mount(
      <Provider store={store}>
        <Input />
      </Provider>
    );

    wrapper.find('ExecuteButton').simulate('click');
    const expected = {
      type: 'appSaga/START_COMMAND_SEQUENCE',
      commands: ['PLACE 1,2,EAST', 'MOVE', 'MOVE', 'LEFT', 'MOVE', 'REPORT'],
    };
    expect(store.getActions()).toEqual([expected]);
  });

  it('renders Input that matches the snapshot', () => {
    const store = mockStore({
      app: { isRunning: false },
    });
    const wrapper = mount(
      <Provider store={store}>
        <Input />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
