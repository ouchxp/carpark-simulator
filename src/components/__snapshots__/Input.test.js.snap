// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders Input that matches the snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Input)>
    <Input
      dispatch={[Function]}
      isRunning={false}
    >
      <label
        htmlFor="commandInput"
      >
        Input:
      </label>
      <InputArea
        defaultValue="PLACE 1,2,EAST
MOVE
MOVE
LEFT
MOVE
REPORT"
        disabled={false}
        id="commandInput"
        placeholder="write one command each line"
      >
        <StyledComponent
          defaultValue="PLACE 1,2,EAST
MOVE
MOVE
LEFT
MOVE
REPORT"
          disabled={false}
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "InputArea-jnFRdt",
                "isStatic": false,
                "lastClassName": "eHYbFz",
                "rules": Array [
                  "font-family:monospace;width:-moz-available;width:-webkit-fill-available;height:140px;resize:none;border:1px solid lightgrey;border-radius:4px;",
                ],
              },
              "displayName": "InputArea",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "InputArea-jnFRdt",
              "target": "textarea",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={
            Object {
              "current": <textarea
                class="InputArea-jnFRdt eHYbFz"
                id="commandInput"
                placeholder="write one command each line"
              >
                PLACE 1,2,EAST
MOVE
MOVE
LEFT
MOVE
REPORT
              </textarea>,
            }
          }
          id="commandInput"
          placeholder="write one command each line"
        >
          <textarea
            className="InputArea-jnFRdt eHYbFz"
            defaultValue="PLACE 1,2,EAST
MOVE
MOVE
LEFT
MOVE
REPORT"
            disabled={false}
            id="commandInput"
            placeholder="write one command each line"
          />
        </StyledComponent>
      </InputArea>
      <ExecuteButton
        disabled={false}
        onClick={[Function]}
      >
        <StyledComponent
          disabled={false}
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "ExecuteButton-zSjZw",
                "isStatic": false,
                "lastClassName": "bUimEV",
                "rules": Array [
                  "display:block;margin-right:0;margin-left:auto;",
                ],
              },
              "displayName": "ExecuteButton",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "ExecuteButton-zSjZw",
              "target": "button",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
          onClick={[Function]}
        >
          <button
            className="ExecuteButton-zSjZw bUimEV"
            disabled={false}
            onClick={[Function]}
          >
            Execute
          </button>
        </StyledComponent>
      </ExecuteButton>
    </Input>
  </Connect(Input)>
</Provider>
`;
